From: =?UTF-8?q?V=C3=A1clav=20Slav=C3=ADk?= <vaclav@slavik.io>
Date: Mon, 15 Apr 2019 18:18:54 +0200
Subject: [PATCH 2/5] Add GETTEXTIOENCODING support

This patch adds support for GETTEXTIOENCODING environment variable that can be
set to override console codepage autodetection. This is particularly useful
when redirecting the output to a pipe (console CP detection fails) and using
localized messages. Inspired by Python's PYTHONIOENCODING.
---
 gettext-runtime/gnulib-lib/localcharset.c | 6 ++++++
 gettext-runtime/intl/localcharset.c       | 6 ++++++
 gettext-tools/gnulib-lib/localcharset.c   | 6 ++++++
 gettext-tools/libgettextpo/localcharset.c | 6 ++++++
 4 files changed, 24 insertions(+)

--- a/gettext-runtime/gnulib-lib/localcharset.c
+++ b/gettext-runtime/gnulib-lib/localcharset.c
@@ -688,6 +688,12 @@ locale_charset (void)
 {
   const char *codeset;
 
+  /* Force output encoding, particularly useful for redirected output on win32
+     where charset detection fails when no actual console is attached. */
+  const char *forced_encoding = getenv("GETTEXTIOENCODING");
+  if (forced_encoding)
+    return forced_encoding;
+
 #if HAVE_LANGINFO_CODESET || defined WINDOWS_NATIVE || defined OS2
 
 # if HAVE_LANGINFO_CODESET
diff --git a/gettext-runtime/intl/localcharset.c b/gettext-runtime/intl/localcharset.c
index e4f47f6..4e89c18 100644
--- a/gettext-runtime/intl/localcharset.c
+++ b/gettext-runtime/intl/localcharset.c
@@ -688,6 +688,12 @@ locale_charset (void)
 {
   const char *codeset;
 
+  /* Force output encoding, particularly useful for redirected output on win32
+     where charset detection fails when no actual console is attached. */
+  const char *forced_encoding = getenv("GETTEXTIOENCODING");
+  if (forced_encoding)
+    return forced_encoding;
+
 #if HAVE_LANGINFO_CODESET || defined WINDOWS_NATIVE || defined OS2
 
 # if HAVE_LANGINFO_CODESET
diff --git a/gettext-tools/gnulib-lib/localcharset.c b/gettext-tools/gnulib-lib/localcharset.c
index 38e27e6..07fd242 100644
--- a/gettext-tools/gnulib-lib/localcharset.c
+++ b/gettext-tools/gnulib-lib/localcharset.c
@@ -688,6 +688,12 @@ locale_charset (void)
 {
   const char *codeset;
 
+  /* Force output encoding, particularly useful for redirected output on win32
+     where charset detection fails when no actual console is attached. */
+  const char *forced_encoding = getenv("GETTEXTIOENCODING");
+  if (forced_encoding)
+    return forced_encoding;
+
 #if HAVE_LANGINFO_CODESET || defined WINDOWS_NATIVE || defined OS2
 
 # if HAVE_LANGINFO_CODESET
diff --git a/gettext-tools/libgettextpo/localcharset.c b/gettext-tools/libgettextpo/localcharset.c
index 38e27e6..07fd242 100644
--- a/gettext-tools/libgettextpo/localcharset.c
+++ b/gettext-tools/libgettextpo/localcharset.c
@@ -688,6 +688,12 @@ locale_charset (void)
 {
   const char *codeset;
 
+  /* Force output encoding, particularly useful for redirected output on win32
+     where charset detection fails when no actual console is attached. */
+  const char *forced_encoding = getenv("GETTEXTIOENCODING");
+  if (forced_encoding)
+    return forced_encoding;
+
 #if HAVE_LANGINFO_CODESET || defined WINDOWS_NATIVE || defined OS2
 
 # if HAVE_LANGINFO_CODESET
-- 
2.20.1.windows.1

